packages/cli/src/render-solidity/renderTablesFromConfig.ts

```diff
@@ -1,47 +1,67 @@
+import path from "path";
 import { renderTable } from "./renderTable.js";
-import { SchemaType, SchemaTypeArrayToElement, SchemaTypeId, getStaticByteLength } from "@latticexyz/schema-type";
+import { SchemaTypeArrayToElement } from "@latticexyz/schema-type";
 import { StoreConfig } from "../config/loadStoreConfig.js";
 import {
+  ImportDatum,
   RenderTableDynamicField,
   RenderTableField,
   RenderTablePrimaryKey,
   RenderTableStaticField,
-  RenderTableType,
 } from "./types.js";
+import { getSchemaTypeInfo, resolveSchemaOrUserType } from "./userType.js";
 
-export function renderTablesFromConfig(config: StoreConfig) {
+export function renderTablesFromConfig(config: StoreConfig, srcDirectory: string) {
   const storeImportPath = config.storeImportPath;
 
   const renderedTables = [];
   for (const tableName of Object.keys(config.tables)) {
     const tableData = config.tables[tableName];
+    const outputDirectory = path.join(srcDirectory, tableData.directory);
 
     // struct adds methods to get/set all values at once
     const withStruct = tableData.dataStruct;
     // operate on all fields at once; for only 1 field keep them only if struct is also kept
     const withRecordMethods = withStruct || Object.keys(tableData.schema).length > 1;
     // field methods can be simply get/set if there's only 1 field and no record methods
     const noFieldMethodSuffix = !withRecordMethods && Object.keys(tableData.schema).length === 1;
+    // list of any symbols that need to be imported
+    const imports: ImportDatum[] = [];
 
     const primaryKeys = Object.keys(tableData.primaryKeys).map((name) => {
-      const type = tableData.primaryKeys[name];
-      const typeInfo = getSchemaTypeInfo(type);
-      if (typeInfo.isDynamic) throw new Error("Parsing error: found dynamic primary key");
+      const schemaOrUserType = tableData.primaryKeys[name];
+      const { renderTableType, importDatum } = resolveSchemaOrUserType(
+        schemaOrUserType,
+        srcDirectory,
+        outputDirectory,
+        config.userTypes
+      );
+      if (importDatum) imports.push(importDatum);
+
+      if (renderTableType.isDynamic) throw new Error("Parsing error: found dynamic primary key");
```

Comment:
added more details
(for context: the errors in `render-solidity` aren't as good because they should never happen unless the user just skips config parsing and calls renderer directly. It also just skips validation in favor of strong typing whenever possible.)